core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatEntry.scala

```diff
@@ -294,80 +331,249 @@ class ColumnFormatValue extends SerializedDiskBuffer
+        val position = buffer.position()
+        val typeId = buffer.getInt(position)
+        if (typeId >= 0) {
+          this.decompressionState = 1
+          return this
+        }
+        // replace underlying buffer if either no other thread is holding a reference
+        // or if this is a heap buffer
+        val allocator = GemFireCacheImpl.getCurrentBufferAllocator
+        val context = this.regionContext
+        val perfStats = getCachePerfStats(context)
+        val startDecompression = perfStats.startDecompression()
+        val decompressed = Utils.codecDecompress(buffer, allocator, position, -typeId)
+        // update decompression stats
+        perfStats.endDecompression(startDecompression)
+        val newValue = copy(decompressed, isCompressed = false, changeOwnerToStorage = false)
+        if (!isDirect || this.refCount <= 2) {
+          val updateStats = (context ne null) && !fromDisk
+          if (updateStats && !isDirect) {
+            StoreCallbacksImpl.acquireStorageMemory(context.getFullPath,
+              decompressed.capacity() - buffer.capacity(), buffer = null,
+              offHeap = false, shouldEvict = true)
+          }
+          val newBuffer = transferToStorage(decompressed, allocator)
+          // update the statistics before changing self
+          if (updateStats) {
+            context.updateMemoryStats(this, newValue)
+          }
+          this.columnBuffer = newBuffer
+          this.decompressionState = 1
+          if (isDirect) {
+            UnsafeHolder.releaseDirectBuffer(buffer)
+          }
+          this
+        } else {
+          perfStats.incDecompressedReplaceSkipped()
+          newValue
+        }
+      }
+    } else if (compress) {
+      // a negative value indicates that the buffer is not compressible (either too
+      // small or minimum compression ratio is not achieved)
+      if (this.decompressionState <= 0) this
+      else {
+        // compress buffer if required
+        if (compressionCodecId != CompressionCodecId.None.id) {
+          val isDirect = buffer.isDirect
+          // check if stored buffer should also be replaced (if no other thread
+          // is holding a reference or this is a heap buffer)
+          val maxCompressionsExceeded = this.decompressionState >
+              ColumnFormatEntry.MAX_CONSECUTIVE_COMPRESSIONS
+          val allocator = GemFireCacheImpl.getCurrentBufferAllocator
+          val context = this.regionContext
+          val perfStats = getCachePerfStats(context)
+          val startCompression = perfStats.startCompression()
+          val bufferLen = buffer.remaining()
+          val compressed = Utils.codecCompress(compressionCodecId, buffer, bufferLen, allocator)
+          if (compressed ne buffer) {
+            // update compression stats
+            perfStats.endCompression(startCompression, bufferLen, compressed.limit())
+            if (maxCompressionsExceeded && (!isDirect || this.refCount <= 2)) {
```

Comment:
Didn't get this. The compressed result replaces the underlying buffer with "this" being returned back. You mean reduce the MAX_CONSECUTIVE_COMPRESSIONS limit? The only extra thing it does before storing is that the compressed buffer should not be wasting too much of unused space at the end because that cannot be tolerated in storage but this trimming is skipped when the buffer is not to be stored. That is some extra transient memory usage to avoid the cost of a copy (which is why it uses runtime pool for compressed/decompressed output in Utils).